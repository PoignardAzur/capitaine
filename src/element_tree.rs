use crate::glue::GlobalEventCx;
use crate::widget_sequence::WidgetSequence;

pub trait ElementTree<ExplicitState> {
    type Event;
    type AggregateComponentState: Default;
    type BuildOutput: VirtualDom<
        ExplicitState,
        Event = Self::Event,
        AggregateComponentState = Self::AggregateComponentState,
    >;

    fn build(
        self,
        prev_state: Self::AggregateComponentState,
    ) -> (Self::BuildOutput, Self::AggregateComponentState);
}

// TODO - Include documentation about what a Virtual DOM is and where the name comes from.
pub trait VirtualDom<ParentComponentState> {
    type Event;
    type AggregateComponentState: Default;

    // TODO - Might be superfluous
    type DomState;

    type TargetWidgetSeq: WidgetSequence;

    // update_value is intended to enable memoize-style HOC
    // where instead of returning a vdom node, it returns
    // something along the lines of struct KeepEverythingAsItWas()
    // Ugh. I'm not explaining this well.
    fn update_value(&mut self, other: Self);

    fn init_tree(&self) -> (Self::TargetWidgetSeq, Self::DomState);

    fn apply_diff(
        &self,
        other: &Self,
        prev_state: Self::DomState,
        widget_seq: &mut Self::TargetWidgetSeq,
    ) -> Self::DomState;

    fn process_event(
        &self,
        explicit_state: &mut ParentComponentState,
        children_state: &mut Self::AggregateComponentState,
        dom_state: &mut Self::DomState,
        _cx: &mut GlobalEventCx,
    ) -> Option<Self::Event>;
}

/*
use crate::glue::{DruidAppData, GlobalEventCx, Id};

use crate::react_widgets::make_button;
use crate::react_widgets::{SingleWidget, WidgetList, WidgetSequence, WidgetTuple};

// TODO - refactor away WidgetPod
use druid::widget as druid_w;
use druid::WidgetPod;


use druid::kurbo::{Rect, Size};

use crate::flex2::FlexParams;
use crate::glue::DruidAppData;
use druid::widget::Button;
use druid::{
    BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, UpdateCtx,
    Widget, WidgetPod,
};
*/
