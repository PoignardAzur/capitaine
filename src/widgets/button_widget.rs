use crate::glue::Action;
use crate::glue::DruidAppData;
use crate::glue::Id;
use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;
use crate::widgets::flex::FlexParams;

use druid::kurbo::{Rect, Size};
use druid::{
    BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle,
    LifeCycleCtx, PaintCtx, UpdateCtx, WidgetPod,
};
use druid::widget::{Button, Click, ControllerHost};

pub struct ButtonWidget(
    pub WidgetPod<DruidAppData, ControllerHost<Button<DruidAppData>, Click<DruidAppData>>>,
    pub Id,
);

impl ButtonWidget {
    pub fn new(text: String, id: Id) -> Self {
        let button = Button::new(text)
            .on_click(move |_, data: &mut DruidAppData, _| data.queue_action(id, Action::Clicked));

        ButtonWidget(WidgetPod::new(button), id)
    }
}

impl FlexWidget for ButtonWidget {
    fn flex_params(&self) -> FlexParams {
        FlexParams {
            flex: 1.0,
            alignment: None,
        }
    }

    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut DruidAppData, env: &Env) {
        self.0.event(ctx, event, data, env);
    }

    fn lifecycle(
        &mut self,
        ctx: &mut LifeCycleCtx,
        event: &LifeCycle,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.lifecycle(ctx, event, data, env);
    }   fn update(
        &mut self,
        ctx: &mut UpdateCtx,
        _old_data: &DruidAppData,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.update(ctx, data, env);
    }

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        bc: &BoxConstraints,
        data: &DruidAppData,
        env: &Env,
    ) -> Size {
        self.0.layout(ctx, bc, data, env)
    }

    fn paint_rect(&self) -> Rect {
        self.0.paint_rect()
    }

    fn set_layout_rect(&mut self, ctx: &mut LayoutCtx, data: &DruidAppData, env: &Env, rect: Rect) {
        self.0.set_layout_rect(ctx, data, env, rect)
    }

    fn layout_rect(&self) -> Rect {
        self.0.layout_rect()
    }

    fn paint(&mut self, ctx: &mut PaintCtx, data: &DruidAppData, env: &Env) {
        self.0.paint(ctx, data, env);
    }
}

impl WidgetSequence for ButtonWidget {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        vec![self]
    }
}
