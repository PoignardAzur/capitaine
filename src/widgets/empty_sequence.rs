use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

pub struct EmptySequence;

impl WidgetSequence for EmptySequence {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        Vec::new()
    }
}
