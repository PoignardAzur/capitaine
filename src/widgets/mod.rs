pub mod empty_sequence;
pub mod flex_widget;
pub mod optional_widget;
pub mod single_widget;
pub mod widget_list;
pub mod widget_tuple;

pub use empty_sequence::*;
pub use flex_widget as flex;
pub use optional_widget::*;
pub use single_widget::*;
pub use widget_list::*;
pub use widget_tuple::*;
