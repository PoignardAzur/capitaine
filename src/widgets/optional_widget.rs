use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

use either::{Either, Left, Right};

// ---

impl<Child: WidgetSequence> WidgetSequence for Option<Child> {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.iter_mut().flat_map(|child| child.widgets()).collect()
    }
}

impl<ChildLeft: WidgetSequence, ChildRight: WidgetSequence> WidgetSequence
    for Either<ChildLeft, ChildRight>
{
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        // TODO - use iterator chain instead?
        match self {
            Left(child_left) => child_left.widgets(),
            Right(child_right) => child_right.widgets(),
        }
    }
}
