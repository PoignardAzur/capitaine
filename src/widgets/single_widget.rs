use crate::glue::DruidAppData;
use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;
use crate::widgets::flex::FlexParams;
use druid::kurbo::{Rect, Size};
use druid::{
    BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, UpdateCtx,
    Widget, WidgetPod,
};

pub struct SingleWidget<W: Widget<DruidAppData>>(pub WidgetPod<DruidAppData, W>);

impl<W: Widget<DruidAppData>> SingleWidget<W> {
    pub fn new(widget: W) -> Self {
        SingleWidget(WidgetPod::new(widget))
    }
}

impl<W: Widget<DruidAppData>> FlexWidget for SingleWidget<W> {
    fn flex_params(&self) -> FlexParams {
        FlexParams {
            flex: 1.0,
            alignment: None,
        }
    }

    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut DruidAppData, env: &Env) {
        self.0.event(ctx, event, data, env);
    }

    fn lifecycle(
        &mut self,
        ctx: &mut LifeCycleCtx,
        event: &LifeCycle,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.lifecycle(ctx, event, data, env);
    }

    fn update(
        &mut self,
        ctx: &mut UpdateCtx,
        _old_data: &DruidAppData,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.update(ctx, data, env);
    }

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        bc: &BoxConstraints,
        data: &DruidAppData,
        env: &Env,
    ) -> Size {
        self.0.layout(ctx, bc, data, env)
    }

    fn paint_rect(&self) -> Rect {
        self.0.paint_rect()
    }

    fn set_layout_rect(&mut self, ctx: &mut LayoutCtx, data: &DruidAppData, env: &Env, rect: Rect) {
        self.0.set_layout_rect(ctx, data, env, rect)
    }

    fn layout_rect(&self) -> Rect {
        self.0.layout_rect()
    }

    fn paint(&mut self, ctx: &mut PaintCtx, data: &DruidAppData, env: &Env) {
        self.0.paint(ctx, data, env);
    }
}

impl<W: Widget<DruidAppData>> WidgetSequence for SingleWidget<W> {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        vec![self]
    }
}

use crate::glue::Action;
use crate::glue::Id;
use druid::widget::Button;
use druid::widget::Click;
use druid::widget::ControllerHost;
pub fn make_button(
    text: String,
    id: Id,
) -> SingleWidget<ControllerHost<Button<DruidAppData>, Click<DruidAppData>>> {
    let button = Button::new(text)
        .on_click(move |_, data: &mut DruidAppData, _| data.queue_action(id, Action::Clicked));

    SingleWidget(WidgetPod::new(button))
}
