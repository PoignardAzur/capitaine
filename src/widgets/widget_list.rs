use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

pub struct WidgetList<Child: WidgetSequence> {
    pub children: Vec<Child>,
}

impl<Child: WidgetSequence> WidgetSequence for WidgetList<Child> {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.children
            .iter_mut()
            .flat_map(|child| child.widgets())
            .collect()
    }
}
