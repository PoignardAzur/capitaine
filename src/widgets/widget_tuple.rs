use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

pub struct WidgetTuple<
    WS0: WidgetSequence,
    WS1: WidgetSequence,
    WS2: WidgetSequence,
    WS3: WidgetSequence,
>(pub WS0, pub WS1, pub WS2, pub WS3);

impl<WS0: WidgetSequence, WS1: WidgetSequence, WS2: WidgetSequence, WS3: WidgetSequence>
    WidgetSequence for WidgetTuple<WS0, WS1, WS2, WS3>
{
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        let mut all_widgets = Vec::new();
        all_widgets.append(&mut self.0.widgets());
        all_widgets.append(&mut self.1.widgets());
        all_widgets.append(&mut self.2.widgets());
        all_widgets.append(&mut self.3.widgets());
        all_widgets
    }
}
