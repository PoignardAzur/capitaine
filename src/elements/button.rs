use crate::glue::{DruidAppData, GlobalEventCx, Id};

use crate::element_tree::{ElementTree, VirtualDom};
use crate::widgets::{make_button, SingleWidget};

use derivative::Derivative;
use druid::widget as druid_w;
use druid::widget::Click;
use druid::widget::ControllerHost;

#[derive(Derivative, Clone, Default, PartialEq, Eq, Hash)]
#[derivative(Debug(bound = ""))]
pub struct Button<ExplicitState>(pub String, pub std::marker::PhantomData<ExplicitState>);

#[derive(Derivative, Clone, Default, PartialEq, Eq, Hash)]
#[derivative(Debug(bound = ""))]
pub struct ButtonData<ParentComponentState>(
    pub String,
    pub std::marker::PhantomData<ParentComponentState>,
);

//
// --- IMPLS

impl<ExplicitState> Button<ExplicitState> {
    pub fn new(text: impl Into<String>) -> Button<ExplicitState> {
        Button(text.into(), Default::default())
    }
}

impl<ExplicitState> ElementTree<ExplicitState> for Button<ExplicitState> {
    type Event = ButtonPressed;
    type AggregateComponentState = ();
    type BuildOutput = ButtonData<ExplicitState>;

    fn build(self, _prev_state: ()) -> (ButtonData<ExplicitState>, ()) {
        (ButtonData(self.0, Default::default()), ())
    }
}

pub struct ButtonPressed();

impl<ParentComponentState> VirtualDom<ParentComponentState> for ButtonData<ParentComponentState> {
    type Event = ButtonPressed;
    type DomState = Id;
    type AggregateComponentState = ();

    // FIXME
    type TargetWidgetSeq =
        SingleWidget<ControllerHost<druid_w::Button<DruidAppData>, Click<DruidAppData>>>;

    fn update_value(&mut self, other: Self) {
        *self = other;
    }

    fn init_tree(&self) -> (Self::TargetWidgetSeq, Id) {
        let text = &self.0;
        let id = Id::new();
        (make_button(text.clone(), id), id)
    }

    fn apply_diff(
        &self,
        _other: &Self,
        prev_state: Self::DomState,
        _widget: &mut Self::TargetWidgetSeq,
    ) -> Id {
        let _text = &self.0;
        //widget.set_text(text.clone());
        prev_state
    }

    fn process_event(
        &self,
        _explicit_state: &mut ParentComponentState,
        _children_state: &mut (),
        dom_state: &mut Id,
        _cx: &mut GlobalEventCx,
    ) -> Option<ButtonPressed> {
        let id = *dom_state;
        if _cx.app_data.dequeue_action(id).is_some() {
            Some(ButtonPressed())
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::element_tree::assign_empty_state_type;
    use insta::assert_debug_snapshot;

    #[test]
    fn new_button() {
        let button = Button::new("Hello");
        let (button_data, ()) = button.clone().build(());

        assert_debug_snapshot!(button);
        assert_debug_snapshot!(button_data);

        assert_eq!(
            button_data,
            ButtonData(String::from("Hello"), Default::default())
        );

        assign_empty_state_type(&button);
    }

    // TODO
    // - Id test (??)
    // - Event test
    // - Widget test
}
