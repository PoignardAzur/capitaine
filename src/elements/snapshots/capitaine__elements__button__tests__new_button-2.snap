---
source: src/elements/button.rs
expression: button_data
---
ButtonData(
    "Hello",
    PhantomData,
)
